File: clients/python/memcore.py
"""A minimal, dependency-free Memcore client speaking RESP2 over TCP.

Memcore is wire-compatible with Redis, so this module also works against a Redis
server, and any existing Redis client (redis-py) works against Memcore. Use this
when a zero-dependency client is preferable to a full-featured one.

Example:
    with Memcore(host="127.0.0.1", port=6380) as c:
        c.set("greeting", "hello")
        print(c.get("greeting"))  # "hello"
"""

from __future__ import annotations

import socket
from typing import Optional, Sequence, Union

Arg = Union[str, int, bytes]
Reply = object  # str | int | bytes | None | list, decoded from RESP2


class MemcoreError(Exception):
    """Raised for a server error reply (a RESP '-ERR ...' line).

    This is distinct from transport failures, which raise ConnectionError or
    OSError.
    """


class Memcore:
    """A synchronous connection to a Memcore server.

    Not safe for concurrent use from multiple threads; give each thread its own
    connection.
    """

    def __init__(self, host: str = "127.0.0.1", port: int = 6380, timeout: float = 5.0):
        self._sock = socket.create_connection((host, port), timeout)
        self._sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        # A buffered reader handles the line-and-length framing of RESP cleanly.
        self._reader = self._sock.makefile("rb")

    def command(self, *args: Arg) -> Reply:
        """Send one command and return its decoded reply."""
        self._sock.sendall(_encode(args))
        return _read_reply(self._reader)

    def get(self, key: str) -> Optional[str]:
        return self.command("GET", key)

    def set(self, key: str, value: Arg) -> Reply:
        return self.command("SET", key, value)

    def delete(self, *keys: str) -> int:
        return self.command("DEL", *keys)

    def ping(self, message: Optional[str] = None) -> Reply:
        return self.command("PING") if message is None else self.command("PING", message)

    def close(self) -> None:
        try:
            self._reader.close()
        finally:
            self._sock.close()

    def __enter__(self) -> "Memcore":
        return self

    def __exit__(self, *_exc) -> None:
        self.close()


def _encode(args: Sequence[Arg]) -> bytes:
    out = bytearray(b"*%d\r\n" % len(args))
    for arg in args:
        b = arg if isinstance(arg, (bytes, bytearray)) else str(arg).encode()
        out += b"$%d\r\n" % len(b)
        out += b
        out += b"\r\n"
    return bytes(out)


def _read_reply(reader) -> Reply:
    line = reader.readline()
    if not line:
        raise ConnectionError("memcore: connection closed")
    kind, payload = line[:1], line[1:-2]  # strip the trailing CRLF

    if kind == b"+":
        return payload.decode()
    if kind == b"-":
        raise MemcoreError(payload.decode())
    if kind == b":":
        return int(payload)
    if kind == b"$":
        length = int(payload)
        if length == -1:
            return None
        data = reader.read(length)
        reader.read(2)  # discard the CRLF after the body
        return data.decode()
    if kind == b"*":
        count = int(payload)
        if count == -1:
            return None
        return [_read_reply(reader) for _ in range(count)]
    raise MemcoreError("memcore: unknown reply type %r" % kind)

File: clients/node/memcore.mjs
// A minimal, dependency-free Memcore client speaking RESP2 over TCP. Memcore is
// wire-compatible with Redis, so this also works against a Redis server, and any
// existing Redis client (ioredis, node-redis) works against Memcore. Use this
// when a zero-dependency client is preferable to a full-featured one.
//
// The reply parser is incremental: Node delivers socket data in arbitrary
// chunks, so a single reply may arrive across several "data" events, and several
// replies may arrive in one. Pending requests are resolved in the order they
// were sent, which is also what makes pipelining work.

import net from "node:net";

// MemcoreError is thrown for a server error reply (a RESP "-ERR ..." line). It
// is distinct from transport errors, which surface as ordinary Error objects.
export class MemcoreError extends Error {
  constructor(message) {
    super(message);
    this.name = "MemcoreError";
  }
}

// connect opens a connection and resolves once it is established.
export function connect({ host = "127.0.0.1", port = 6380 } = {}) {
  const client = new Client(host, port);
  return client._connect().then(() => client);
}

class Client {
  constructor(host, port) {
    this.host = host;
    this.port = port;
    this.socket = null;
    this._pending = []; // queued {resolve, reject}, one per in-flight request
    this._buffer = Buffer.alloc(0);
  }

  _connect() {
    return new Promise((resolve, reject) => {
      this.socket = net.createConnection({ host: this.host, port: this.port });
      this.socket.setNoDelay(true);
      const onError = (err) => reject(err);
      this.socket.once("error", onError);
      this.socket.once("connect", () => {
        this.socket.removeListener("error", onError);
        this.socket.on("data", (chunk) => this._onData(chunk));
        this.socket.on("error", (err) => this._fail(err));
        this.socket.on("close", () => this._fail(new Error("connection closed")));
        resolve();
      });
    });
  }

  // command sends a command and resolves with its decoded reply. Arguments are
  // coerced to strings; pass a Buffer for binary-safe values.
  command(...args) {
    return new Promise((resolve, reject) => {
      this._pending.push({ resolve, reject });
      this.socket.write(encodeCommand(args));
    });
  }

  get(key) {
    return this.command("GET", key);
  }
  set(key, value) {
    return this.command("SET", key, value);
  }
  del(...keys) {
    return this.command("DEL", ...keys);
  }
  ping(message) {
    return message === undefined ? this.command("PING") : this.command("PING", message);
  }

  close() {
    if (this.socket) this.socket.end();
  }

  _onData(chunk) {
    this._buffer = Buffer.concat([this._buffer, chunk]);
    for (;;) {
      const parsed = parseReply(this._buffer, 0);
      if (parsed === null) break; // a complete reply is not yet buffered
      this._buffer = this._buffer.subarray(parsed.offset);
      const request = this._pending.shift();
      if (!request) continue; // reply with no waiter; ignore defensively
      if (parsed.value instanceof MemcoreError) request.reject(parsed.value);
      else request.resolve(parsed.value);
    }
  }

  _fail(err) {
    while (this._pending.length) this._pending.shift().reject(err);
  }
}

function encodeCommand(args) {
  const parts = [Buffer.from(`*${args.length}\r\n`)];
  for (const arg of args) {
    const buf = Buffer.isBuffer(arg) ? arg : Buffer.from(String(arg));
    parts.push(Buffer.from(`$${buf.length}\r\n`), buf, Buffer.from("\r\n"));
  }
  return Buffer.concat(parts);
}

// parseReply decodes one reply starting at offset. It returns { value, offset }
// where offset is the index just past the reply, or null when the buffer does
// not yet hold a complete reply.
function parseReply(buf, offset) {
  if (offset >= buf.length) return null;
  const crlf = buf.indexOf("\r\n", offset + 1);
  if (crlf === -1) return null;

  const type = buf[offset];
  const line = buf.toString("utf8", offset + 1, crlf);
  const afterLine = crlf + 2;

  switch (type) {
    case 0x2b: // '+' simple string
      return { value: line, offset: afterLine };
    case 0x2d: // '-' error
      return { value: new MemcoreError(line), offset: afterLine };
    case 0x3a: // ':' integer
      return { value: Number(line), offset: afterLine };
    case 0x24: { // '$' bulk string
      const length = Number(line);
      if (length === -1) return { value: null, offset: afterLine };
      const end = afterLine + length;
      if (buf.length < end + 2) return null; // body plus its CRLF not all here
      return { value: buf.toString("utf8", afterLine, end), offset: end + 2 };
    }
    case 0x2a: { // '*' array
      const count = Number(line);
      if (count === -1) return { value: null, offset: afterLine };
      const items = [];
      let cursor = afterLine;
      for (let i = 0; i < count; i++) {
        const element = parseReply(buf, cursor);
        if (element === null) return null;
        items.push(element.value);
        cursor = element.offset;
      }
      return { value: items, offset: cursor };
    }
    default:
      throw new Error(`memcore: unknown reply type ${String.fromCharCode(type)}`);
  }
}
